apis/kernel/ipc/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+#![no_std]
+
+use libtock_platform as platform;
+use platform::{
+    allow_rw, exit_on_drop, return_variant, share, subscribe, syscall_class, DefaultConfig,
+    ErrorCode, Register, ReturnVariant, Syscalls, Upcall,
+};
+
+/// The IPC driver.
+///
+/// # Example
+///
+/// Service:
+///
+/// ```ignore
+/// use libtock::ipc::{Ipc, IpcCallData, IpcListener};
+/// use libtock::leds::Leds;
+///
+/// fn led_callback(data: IpcCallData) {
+///     let _ = Leds::on(0);
+/// }
+///
+/// // Creates an IPC service for turning on an LED
+/// let listener = IpcListener(led_callback);
+///
+/// // Registers the IPC service
+/// let _ = Ipc::register_service_listener(listener);
+/// ```
+///
+/// Client:
+///
+/// ```ignore
+/// use libtock::ipc::Ipc;
+///
+/// // Discovers the IPC service
+/// let service_id = Ipc::discover("org.tockos.example.led").unwrap();
+///
+/// // Runs the IPC service
+/// let _ = Ipc::notify_service(service_id);
+/// ````
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct IpcCallData<'a> {
+    pub caller_id: u32,
+    pub buffer: Option<&'a mut [u8]>,
```

Comment:
Should we consider the lifetime of this slice to be `'static`?

My thoughts are that when one process shares a buffer with another via `Ipc::share`, it's effectively handing ownership over to the target processes.  The sharing process needs some way of knowing when the target is done with the buffer and is handing it back.

Example:
A shares `buf` with B.
A notifies B.
B upcall is invoked with (A::id, buf).  B now owns buf.
B performs some long running calculation on buf.
B notifies A.
A upcall is invoked with (B::id, None). A now owns buf again.

Reply:
Ooh, I think that's not a bad idea. I only instantiate it as `IpcCallData<'static>` in the userspace driver itself, but I think for extra clarity removing the generic lifetime and replacing it with `'static` makes sense. Will also add this with the above issue.